robotoff/workers/tasks/import_image.py

```diff
@@ -47,37 +68,95 @@ def run_import_image_job(
-            save_image(barcode, source_image, product, server_domain)
-            import_insights_from_image(
-                barcode, image, source_image, ocr_url, server_domain
-            )
-        with db.atomic():
-            # Launch object detection in a new SQL transaction
-            run_object_detection(barcode, image, source_image, server_domain)
+        save_image(barcode, source_image, product, server_domain)
+
+    enqueue_job(
+        import_insights_from_image,
+        high_queue,
+        barcode=barcode,
+        image_url=image_url,
+        ocr_url=ocr_url,
+        server_domain=server_domain,
+    )
+    enqueue_job(
+        run_logo_object_detection,
+        high_queue,
+        barcode=barcode,
+        image_url=image_url,
+        server_domain=server_domain,
+    )
+    enqueue_job(
+        run_nutrition_table_object_detection,
+        high_queue,
+        barcode=barcode,
+        image_url=image_url,
+        server_domain=server_domain,
+    )
 
 
 def import_insights_from_image(
     barcode: str,
-    image: Image.Image,
-    source_image: str,
+    image_url: str,
     ocr_url: str,
     server_domain: str,
 ):
-    predictions_all = get_predictions_from_image(barcode, image, source_image, ocr_url)
+    image = get_image_from_url(image_url, error_raise=False, session=http_session)
+
+    if image is None:
+        logger.info(f"Error while downloading image {image_url}")
+        return
+
+    source_image = get_source_from_url(image_url)
+    predictions = extract_ocr_predictions(
+        barcode, ocr_url, DEFAULT_OCR_PREDICTION_TYPES
+    )
+    if any(
+        prediction.value_tag == "en:nutriscore"
+        and prediction.type == PredictionType.label
+        for prediction in predictions
+    ):
+        enqueue_job(
+            run_nutriscore_object_detection,
+            high_queue,
+            barcode=barcode,
+            image_url=image_url,
+            server_domain=server_domain,
+        )
     NotifierFactory.get_notifier().notify_image_flag(
-        [p for p in predictions_all if p.type == PredictionType.image_flag],
+        [p for p in predictions if p.type == PredictionType.image_flag],
         source_image,
         barcode,
     )
-    imported = import_insights(predictions_all, server_domain)
-    logger.info(f"Import finished, {imported} insights imported")
+
+    with db:
+        imported = import_insights(predictions, server_domain)
+        logger.info(f"Import finished, {imported} insights imported")
+
+
+@with_db
+def save_image_job(batch: list[tuple[str, str]], server_domain: str):
+    """Save a batch of images in DB.
+
+    :param batch: a batch of (barcode, source_image) tuples
+    :param server_domain: the server domain to use
+    """
+    for barcode, source_image in batch:
+        product = get_product_store()[barcode]
+        if product is None:
+            continue
+        save_image(barcode, source_image, product, server_domain)
```

Comment:
Do you have really big batches ?
If yes, I would really try to fetch existing image in one request, then process the addition in one go (bulk insert).
If you do this one by one, appart that it takes a longer time, you augment the risk of having the whole transaction fail because one image added in between…

Another option to keep code simple, is to use a transaction by save_image (instead of setting it on whole operation).

Reply:
I will switch to a transaction in `save_image`. The idea of this function is to have a simple way to import all images that were not imported due to some failure (see `import_images_in_db` CLI command)